inko/src/pkg/sync.rs

```diff
@@ -0,0 +1,247 @@
+        format!(
+            "The version {} of package {} doesn't exist",
+            dependency.version, url
+        )
+    })?;
+
+    repo.checkout(&tag.target).map_err(|err| {
+        format!(
+            "Failed to checkout tag {} of package {}: {}",
+            tag_name, url, err
+        )
+    })?;
+
+    if tag.target != dependency.checksum.to_string() {
+        format!(
+            "The checksum of {} version {} didn't match.
+
+The checksum that is expected is:
+
+    {}
+
+The actual checksum is:
+
+    {}
+
+This means that either your checksum is incorrect, or the version's contents
+have changed since it was last published.
+
+If the version's contents have changed you'll need to check with the package's
+maintainer to ensure this is expected.
+
+DO NOT PROCEED BLINDLY, as you may be including unexpected or even malicious
+changes.",
+            url, dependency.version, dependency.checksum, tag.target
+        );
+    }
+
+    let package = Package { repository: repo, dependency };
+    let manifest_path = dir.join(MANIFEST_FILE);
+
+    if manifest_path.is_file() {
+        Ok((package, Some(Manifest::load(&manifest_path)?)))
+    } else {
+        Ok((package, None))
+    }
+}
+
+fn remove_dependencies(directory: &Path) -> Result<(), String> {
+    if directory.is_dir() {
+        remove_dir_all(directory).map_err(|err| {
+            format!("Failed to remove {}: {}", directory.display(), err)
+        })?;
+    }
+
+    Ok(())
+}
+
+fn install_packages(
+    packages: Vec<Package>,
+    versions: Vec<(Url, Version)>,
+    directory: &Path,
+) -> Result<(), String> {
+    let repos = packages
+        .into_iter()
+        .map(|pkg| (pkg.dependency.url, pkg.repository))
+        .collect::<HashMap<_, _>>();
+
+    for (url, ver) in versions {
+        let repo = repos.get(&url).unwrap();
+        let tag_name = ver.tag_name();
+        let tag = repo.tag(&tag_name).unwrap();
+
+        repo.checkout(&tag.target).map_err(|err| {
+            format!("Failed to check out {}: {}", tag_name, err)
+        })?;
+
+        let base_dir = directory
+            .join(url.directory_name())
+            .join(&format!("v{}", ver.major));
+
+        let manifest_src = repo.path.join(MANIFEST_FILE);
+
+        cp_r(&repo.path.join(SRC_DIR), &base_dir.join(SRC_DIR))?;
+
+        if manifest_src.is_file() {
+            copy(&manifest_src, &base_dir.join(MANIFEST_FILE))
+                .map_err(|e| format!("Failed to copy inko.pkg: {}", e))?;
+        }
+    }
+
+    Ok(())
+}
+
+fn manifest_hash() -> Result<String, String> {
+    let data = read(MANIFEST_FILE).map_err(|err| {
+        format!(
+            "Failed to read the manifest file at '{}': {}",
+            MANIFEST_FILE, err
+        )
+    })?;
+
+    Ok(blake3::hash(&data).to_hex().to_string())
+}
+
+fn save_manifest_hash(directory: &Path) -> Result<(), String> {
+    if !Path::new(MANIFEST_FILE).exists() {
+        return Ok(());
+    }
```

Comment:
I think this check isn't needed. This method is only called in `sync()`, which in turn is only called if either `inko.pkg` changed (meaning it must exist in the first place), or if one runs `inko pkg sync`.

I think what we can do is the following:

We change `manifest_hash` such that if `read()` returns an error, we check if that error is `std::io::ErrorKind::NotFound`. If so, we return `Ok(None)`. For any other error we return `Err(String)`. If the file does exist, we return `Ok(Some(the hash))`.

In `save_manifest_hash` we then call `manifest_hash`, and if it returns `Ok(None)` we just return early. This way we don't need the extra `exists()` check.